File: Python/0238_productExceptSelf.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # Initialize output list and fill with 1s to the same length as nums
        output = [1] * len(nums)
        # Prefix product is the product of all the values that came before the current index
        prefix = 1
        # Move forwards and calculate
        for i in range(len(nums)):
            # Place prefix product in the output list for later
            output[i] = prefix
            prefix *= nums[i]
        # Postfix product is the product of all the values that came after the current index
        postfix = 1
        # Move backwards and calculate the postfix product
        for i in range(len(nums) - 1, -1, -1):
            # Multiply the prefix by the postfix to calculate the final product
            output[i] *= postfix
            postfix *= nums[i]
        return output

File: Java/0238_productExceptSelf.java
class productExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        // Create a result array with the same amount of elements as the passed nums array
        int[] result = new int[nums.length];
        int prefix = 1;
        // Loop forward through the array
        for (int i = 0; i < nums.length; i++) {
            // Place the current prefix product in the corresponding position in the result array
            result[i] = prefix;
            // Update prefix product with the current number
            prefix *= nums[i];
        }
        int postfix = 1;
        // Loop through the array in reverse
        for (int i = nums.length - 1; i > -1; i--) {
            // Multiply the value in the result array with the postfix product
            result[i] *= postfix;
            // Update the postfix product with the current number
            postfix *= nums[i];
        }
        return result;
    }
}